tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -359,6 +360,182 @@ public function removeRedundantClassesMinifiesClassAttributes($html, array $clas
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string $css
+     * @param string[] $classesExpectedToBeRemoved
+     *
+     * @dataProvider provideClassesNotInUninlinableRules
+     */
+    public function removeRedundantClassesAfterCssInlinedRemovesClassesNotInUninlinableRules(
+        $html,
+        $css,
+        array $classesExpectedToBeRemoved = []
+    ) {
+        \extract($this->buildSubjectAndCssInlinerWithCssInlined('<html>' . $html . '</html>', $css));
+
+        $subject->removeRedundantClassesAfterCssInlined($cssInliner);
+
+        $result = $subject->render();
+        foreach ($classesExpectedToBeRemoved as $class) {
+            self::assertNotContains($class, $result);
+        }
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function provideClassesInUninlinableRules()
+    {
+        return [
+            'media rule' => [
+                'HTML' => '<p class="foo">hello</p>',
+                'CSS' => '@media (max-width: 640px) { .foo { color: green; } }',
+                'classes to be kept' => ['foo'],
+            ],
+            'dynamic pseudo-class' => [
+                'HTML' => '<p class="foo">hello</p>',
+                'CSS' => '.foo:hover { color: green; }',
+                'classes to be kept' => ['foo'],
+            ],
+            '2 classes, in different uninlinable rules' => [
+                'HTML' => '<p class="foo bar">hello</p>',
+                'CSS' => '.foo:hover { color: green; } @media (max-width: 640px) { .bar { color: green; } }',
+                'classes to be kept' => ['foo', 'bar'],
+            ],
+            '1 class in uninlinable rule, 1 in inlinable rule' => [
+                'HTML' => '<p class="foo bar">hello</p>',
+                'CSS' => '.foo { color: red; } .bar:hover { color: green; }',
+                'classes to be kept' => ['bar'],
+            ],
+            '2 classes in same selector' => [
+                'HTML' => '<p class="foo bar">hello</p>',
+                'CSS' => '.foo.bar:hover { color: green; }',
+                'classes to be kept' => ['foo', 'bar'],
+            ],
+            'class with hyphen, underscore, uppercase letter and number in name' => [
+                'HTML' => '<p class="foo-2_A">hello</p>',
+                'CSS' => '.foo-2_A:hover { color: green; }',
+                'classes to be kept' => ['foo-2_A'],
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string $css
+     * @param string[] $classesToBeKept
+     *
+     * @dataProvider provideClassesInUninlinableRules
+     */
+    public function removeRedundantClassesAfterCssInlinedNotRemovesClassesInUninlinableRules(
+        $html,
+        $css,
+        array $classesToBeKept = []
+    ) {
+        \extract($this->buildSubjectAndCssInlinerWithCssInlined('<html>' . $html . '</html>', $css));
```

Comment:
Please let's not use `extract` and `compact` as static code analyzers have a really hard time with these functions. Instead, please let's use `list` and explicit arrays.

Reply:
I was wondering what you'd think about that and aware that using `list` was an alternative solution to the 'multiple return values problem'.  Will change accordingly...